### ISSUE TITLE
KeyDown events trigger the wrong vim key

### ISSUE THREAD

**Issue by romgrk** [rhysd/neovim-component#14]
The failing code is https://github.com/rhysd/neovim-component/blob/master/src/neovim/input.ts#L119
```typescript
vim_input += (special_char || String.fromCharCode(event.keyCode).toLowerCase()) + '>';
```
Example:
Pressing `alt` + `;` generates `<A-º>` where neovim usually gets `<A-;>`
![screenshot from 2016-01-11 15-26-47](https://cloud.githubusercontent.com/assets/1423607/12245637/ca3d1242-b877-11e5-852f-ed60594e57a2.png)

The input is off for all non-letter keys. I recall discussing something like this in https://github.com/rhysd/neovim-component/issues/6, and again I wouldnt know exactly how this could be solved. The problem here is that `String.fromCharCode` takes a character code as argument, and what is being fed to it is a `keyCode`. I had already run in this exact problem when implementing a package for Atom, the only workaround I found then was to have a map of `keycodes`: `characters`, but there was no check of keyboard layout whatsoever, US layout was assumed all the way. That causes evident problems for i18n.
I'll check if there is any possibility to retrieve the user keyboard layout.

**Comment by romgrk** [comment]
This is interesting:
https://github.com/Microsoft/node-native-keymap

**Comment by rhysd** [comment]
Thank you for your report.  It must be a bug on catching user input.

Currently, `<neovim-editor>` catches user input by 2 ways.

- `keydown` event
- `input` event of `<input>` on cursor

`input` is used to catch normal character without modifiers.  I use this because of IM (Input Method).

`keydown` is used for special characters and normal characters with modifiers.  It requires to gain a character string from `keyCode`.  It's difficult because key code is very complicated.  I want to use [`code`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code) but latest Electron (Chromium 47) doesn't implement yet.  (It seems that Chrome 48 beta implements it).

So hopefully `KeyboardEvent.code` will solve this character code related issue.

**Comment by romgrk** [comment]
Looks like it was merged 6 days ago :)
https://github.com/atom/electron/pull/4014
https://github.com/atom/electron/commit/ae6c2d46b89a351110cd1b1bd73248b227adb6c1
Although I'm not sure if the features are enable by default. I'll run some tests.

**Comment by romgrk** [comment]
It works :D (electron 0.36.3)

**Comment by rhysd** [comment]
I updated key handling with `KeyboardEvent.key`.  Finally I added a workaround on `<A-x>` input for OS X.

![works-on-my-machine-starburst](https://cloud.githubusercontent.com/assets/823277/12371830/0dc64848-bc84-11e5-90b0-4fdd181e122b.jpg)

**Comment by romgrk** [comment]
Doesn't seem to be working here (on linux).
I'll run some tests first, but `Alt` seems to be never detected.
Also on some cases `"Unidentified"` is inserted literally to neovim. (which is the value of `KbEvent.key`)

**Comment by rhysd** [comment]
Thank you for confirming.
Hmm... Could you take a screenshot of console which shows the `keydown` event on the case?  

It's weird that `"Unidentified"` is inserted because it should be ignored [here](https://github.com/rhysd/neovim-component/blob/5c880834ebe29f5e06adc2966e3055b93e734467/src/neovim/input.ts#L132).  Anyway, I'll install Ubuntu 14.04 to VirtualBox on my machine and check it.

**Comment by romgrk** [comment]
Yes, indeed. (and I am well on commit 5c88083)
Just a sec, I'll generate the events and post a screenshot. (meanwhile, I was wondering, do you have a ctags parsing tool for typescript?)

**Comment by romgrk** [comment]
`Alt-o`:
![screenshot from 2016-01-17 01-24-56](https://cloud.githubusercontent.com/assets/1423607/12376279/33713c44-bcb9-11e5-86c5-775761612aef.png)

**Comment by romgrk** [comment]
`Ctrl-semicolon`:
![screenshot from 2016-01-17 01-26-36](https://cloud.githubusercontent.com/assets/1423607/12376288/6f09af52-bcb9-11e5-97f4-a21716ecb594.png)

**Comment by romgrk** [comment]
If of any utility, result of tests:
![screenshot from 2016-01-17 01-29-06](https://cloud.githubusercontent.com/assets/1423607/12376296/c122b27a-bcb9-11e5-90ea-ff207ff8c41d.png)

**Comment by rhysd** [comment]
Yes, some tests are not fixed yet, sorry.

**Comment by rhysd** [comment]
(seeing screenshots)

Oh, ctrl key is considered in `KeyboardEvent.key`, but alt key is NOT considered in it...  I didn't know that.  I'll fix it.  Thank you.

**Comment by romgrk** [comment]
Wooo. Getting some really strange output.
When pressing `alt-;`, the `event.key == ";"`
When pressing `ctrl-;` the `event.key == "Unidentified"`

And btw I took a look at the code and what seems to be wrong is this:
https://github.com/rhysd/neovim-component/blob/master/src/neovim/input.ts#L141

I don't know the code like you do, but from what I got it's just a matter of testing if `altKey` or `ctrlKey` are true. (And I do not know if it applies in some cases, but maybe `metaKey`).

**Comment by rhysd** [comment]
I installed Ubuntu 15.10 in virtual environment and I could reproduce what you had reported.

1.  We must input `<C-{key}>` instead of `{key}` when `altKey` is `true`.
2. As you said, `Ctrl+;` is not the same behavior as others.  The keycode was `186` and it's not for `;`.  In addition, `Keyboard.key` is `"Unidentified"` as you reported.  I think it is good to make `Ctrl+;` fall back to `;`.  If we ignore `Ctrl+;` on `keydown` event, succeeding `input` event can catch it and handle input to `<input>` element.  At least, "Unidentified" input to neovim is no good :sob:

**Comment by rhysd** [comment]
Both MacVim and gVim seem to handle `<C-;>` as `;`.  But browser can't know `;` on `<C-;>` because key code in event is wrong in Linux.  So currently `<neovim-editor>` ignores the sequence.  I think it's acceptable.

**Comment by romgrk** [comment]
Yea, I have a local workaround but it would definitely not be clean enough to push. I guess it'll wait until chromium source fixes this. 
I'll be watching upstream.

**Comment by rhysd** [comment]
Thank you.  I added 2 commits to fix them.

1. Ignore `Unidentified` input on 'keydown' event
2. Care alt key (currently e.g. Alt+a should work) also in Linux

**Comment by rhysd** [comment]
This fix is already added to NyaoVim(v0.0.11) and I've detected no issue.  I close this because it seems to work.
